File: legacy/native/macos/Sources/TerraneHostMac/PlatformNotifications.swift
import Foundation

final class PlatformNotifications {
    func toast(_ request: BridgeRequest) -> BridgeResponse {
        guard let message = request.params["message"] as? String else {
            return .failure(id: request.id, code: "invalid_request", message: "notification.toast requires message")
        }
        if let levelValue = request.params["level"], !(levelValue is NSNull) {
            guard let level = levelValue as? String else {
                return .failure(id: request.id, code: "invalid_request", message: "notification.toast level must be a string")
            }
            guard Self.validNotificationLevel(level) else {
                return .failure(
                    id: request.id,
                    code: "invalid_request",
                    message: "notification.toast level must be info, success, warning, or error",
                    details: ["level": level]
                )
            }
        }
        NSLog("Toast: \(message)")
        return .success(id: request.id, result: ["ok": true])
    }

    private static func validNotificationLevel(_ level: String) -> Bool {
        ["info", "success", "warning", "error"].contains(level)
    }
}

File: legacy/native/windows/src/PlatformNotifications.cpp
#include "PlatformNotifications.h"

namespace terrane {
namespace json = winrt::Windows::Data::Json;

namespace {

bool TryGetStringMember(json::JsonObject const& object, std::wstring const& member, std::wstring& out) {
  if (!object.HasKey(member)) {
    return false;
  }
  auto value = object.GetNamedValue(member);
  if (value.ValueType() != json::JsonValueType::String) {
    return false;
  }
  out = value.GetString().c_str();
  return true;
}

bool ValidNotificationLevel(std::wstring const& level) {
  return level == L"info" || level == L"success" || level == L"warning" || level == L"error";
}

}  // namespace

json::JsonObject PlatformNotifications::Toast(BridgeRequest const& request) {
  std::wstring message;
  if (!TryGetStringMember(request.params, L"message", message)) {
    return BridgeResponse::Failure(request.id, request.hasId, L"invalid_request", L"notification.toast requires message");
  }

  if (request.params.HasKey(L"level")) {
    auto levelValue = request.params.GetNamedValue(L"level");
    if (levelValue.ValueType() != json::JsonValueType::Null) {
      std::wstring level;
      if (!TryGetStringMember(request.params, L"level", level)) {
        return BridgeResponse::Failure(
            request.id,
            request.hasId,
            L"invalid_request",
            L"notification.toast level must be a string");
      }
      if (!ValidNotificationLevel(level)) {
        json::JsonObject details;
        details.Insert(L"level", json::JsonValue::CreateStringValue(level));
        return BridgeResponse::Failure(
            request.id,
            request.hasId,
            L"invalid_request",
            L"notification.toast level must be info, success, warning, or error",
            details);
      }
    }
  }

  json::JsonObject result;
  result.Insert(L"ok", json::JsonValue::CreateBooleanValue(true));
  return BridgeResponse::Success(request.id, request.hasId, result);
}

}  // namespace terrane
